.circleci/config.yml

```diff
@@ -1,14 +1,111 @@
+#
+#######
+#
+# As can be seen from the above numbers restoring from workspaces are
+# significantly slower than caches, a fair bit slower than github checkouts and
+# a little bit slower than downloading and installing go.
+#
+# Since workspaces are scoped to a workflow you need to persist to a workspace
+# in your workflow before you can read from it, which further slows the build
+# time and persisting seems to take about twice as long as retrieving. And on
+# top of that remember that before persisting you will need to run the task
+# that you would otherwise run directly in a downstream job.
+#
+# Workspaces work best for small amounts of data that take a long time to
+# generate that we want to share between many jobs without having to duplicate
+# everywhere the logic to generate the data.
+#
+# How are we using them?
+#
+# We are currently using workspaces to store the compiled-system-contracts
+# (4.7M) which take about 5 minutes to generate 2s to persist to the workspace
+# and 0s to restore from the workspace, we also cache them so that most of the
+# time we can skip the build step and simply persist the cache to the
+# workspace. Having this functionality contained in one job makes the config
+# simpler than if we tried to add compiled system contract generation to all of
+# the downstream jobs and since this job is quite fast and not in the critical
+# path for the workflow it's ok to be a bit slower here.
+#
+########
+#
+# The structure of this workflow
+#
+# We have a central folder ~/repos under which there are paths which are used
+# throughout the workflow:
+#
+# - geth: The celo-blockchain checkout
+# - celo-monorepo: The celo-monorepo checkout
+# - golang: A golang installation
+# - gocache: A folder containing intermediate compilation results for golang
+# - gomodcache: A folder containing downloaded modules for golang
 version: 2.1
 parameters:
-  # Increment this to force cache rebuilding
+  # Increment these to force cache rebuilding
   system-contracts-cache-version:
     type: integer
-    default: 1
+    default: 2
+  go-mod-cache-version:
+    type: integer
+    default: 3
+  go-cache-version:
+    type: integer
+    default: 3
+  checkout-monorepo-cache-version:
+    type: integer
+    default: 3
+  # Location where compiled system contracts are stored under the root of this
+  # repo.
   system-contracts-path:
     type: string
     default: "compiled-system-contracts"
   system-contracts-executor-image:
-    type: string
+    type:
+      string
+      # The system contracts executor currently needs node 12 to function,
+      # this should only be changed when the version in `monorepo_commit` is
+      # changed.
+    default: "us.gcr.io/celo-testnet/circleci-node12:1.0.0"
+  e2e-executor-image:
+    type:
+      string
       # The system contracts executor currently needs node 12 to function,
       # this should only be changed when the version in `monorepo_commit` is
       # changed.
```

Comment:
Modify this?

Reply:
Im not sure what you are suggesting here @YoduYodu ?